horreum-backend/src/test/java/io/hyperfoil/tools/horreum/svc/RunServiceTest.java

```diff
@@ -58,32 +66,222 @@ public class RunServiceTest extends BaseServiceTest {
+      BlockingQueue<DataSet> dataSetQueue = eventConsumerQueue(DataSet.class, DataSet.EVENT_NEW);
+      Schema acmeSchema = createExampleSchema("AcmeCorp", "AcmeInc", "AcmeRrUs", false);
+      Schema roadRunnerSchema = createExampleSchema("RoadRunnerCorp", "RoadRunnerInc", "RoadRunnerRrUs", false);
+
+      NamedJsonPath acmePath = createExampleNamedJsonPath("foo", "$.value");
+      String acmeFunction = createFunction("foo");
+      Transformer acmeTransformer = createExampleTransformerWithJsonPath("acme", acmeSchema, acmeFunction, acmePath);
+      NamedJsonPath roadRunnerPath = createExampleNamedJsonPath("bah", "$.value");
+      String bahFunction = createFunction("bah");
+      Transformer roadRunnerTransformer = createExampleTransformerWithJsonPath("roadrunner", roadRunnerSchema, bahFunction, roadRunnerPath);
+
+      Test exampleTest = createExampleTest(getTestName(info));
+      Test test = createTest(exampleTest);
+      addTransformer(test, acmeTransformer, roadRunnerTransformer);
+
+      dataSetQueue.drainTo(new ArrayList<>());
+      int runId = uploadRun(runWithValue(42.0d, acmeSchema, roadRunnerSchema), test.name);
+
+      DataSet event = dataSetQueue.poll(POLL_DURATION_SECONDS, TimeUnit.SECONDS);
+
+      assertNotNull(event);
+      JsonNode node = event.data;
+      assertTrue(node.isArray());
+      assertEquals(2 , node.size());
+      JsonNode acme = node.path(0).path("outcome").path("foo");
+      validate( "42", acme);
+      JsonNode runner = node.path(1).path("outcome").path("bah");
+      validate( "42", runner);
+      try (CloseMe ignored = roleManager.withRoles(em, Arrays.asList(TESTER_ROLES[2]))) {
+         Run run = Run.findById(runId);
+         assertEquals(1, run.datasets.size());
       }
    }
 
+   @org.junit.jupiter.api.Test
+   public void testTransformationSingleSchemaTestWithoutTransformer(TestInfo info) throws InterruptedException {
+      BlockingQueue<DataSet> dataSetQueue = eventConsumerQueue(DataSet.class, DataSet.EVENT_NEW);
+      Schema acmeSchema = createExampleSchema("AceCorp", "AceInc", "AceRrUs", false);
+
+      Test exampleTest = createExampleTest(getTestName(info));
+      Test test = createTest(exampleTest);
+
+      dataSetQueue.drainTo(new ArrayList<>());
+      uploadRun(runWithValue(42.0d, acmeSchema), test.name);
+
+      DataSet event = dataSetQueue.poll(POLL_DURATION_SECONDS, TimeUnit.SECONDS);
+
+      assertNotNull(event);
+      JsonNode node = event.data;
+      assertTrue(node.isArray());
+      ObjectNode object = (ObjectNode)node.path(0);
+      JsonNode schema = object.path("$schema");
+      assertEquals("urn:AceInc:AceRrUs:1.0", schema.textValue());
+      JsonNode value = object.path("value");
+      assertEquals(42, value.intValue());
+   }
+
+   @org.junit.jupiter.api.Test
+   public void testTransformationUsingSameSchemaInBothLevelsTestWithoutTransformer(TestInfo info) throws InterruptedException {
```

Comment:
I think that the output of this test is not correct: in https://github.com/Hyperfoil/Horreum/pull/160/files#r847014832 I've tried to explain why I think this input should produce 2 datasets. Would you like to discuss that further?

Reply:
2 DataSets are now produced as expected.

Reply:
No, your test does not verify that 2 DataSets are produced. This produces only 1 DataSet and verify that it's data is an array of 2 objects with the same schema.

Reply:
Btw. it would be usefuly to not use value 42 for both DataSets - use different values in each schema instance and assert that the datasets contain both values.

Reply:
I agree completely. This will be updated. The tests will be updated also to check the outcome is as expected.